File: 1046-last-stone-weight/1046-last-stone-weight.cpp
class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        make_heap(stones.begin(), stones.end());
        while(stones.size()>1){
            int x=0;
            for(int i=0;i<2;i++){
                pop_heap(stones.begin(), stones.end());
                x = abs(x-stones.back());
                stones.pop_back();
            }
            if(x!=0){
                stones.push_back(x);
                push_heap(stones.begin(), stones.end());
            }
        }
        if(stones.size()>0) return stones.front();
        else return 0;
    }
};

File: 1046-last-stone-weight/1046-last-stone-weight.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        heapq._heapify_max(stones)
        while len(stones) > 1:
            rock1 = heapq._heappop_max(stones)
            rock2 = heapq._heappop_max(stones)
            if rock1 != rock2:
                stones.append(rock1 - rock2)
                heapq._siftdown_max(stones, 0, len(stones)-1)
            
        return sum(stones)